core/src/main/java/bisq/core/app/AppStartupState.java

```diff
@@ -0,0 +1,149 @@
+ *
+ * Bisq is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with Bisq. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package bisq.core.app;
+
+import bisq.core.btc.setup.WalletsSetup;
+
+import bisq.network.p2p.BootstrapListener;
+import bisq.network.p2p.P2PService;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.fxmisc.easybind.EasyBind;
+import org.fxmisc.easybind.monadic.MonadicBinding;
+
+import javafx.beans.property.BooleanProperty;
+import javafx.beans.property.ReadOnlyBooleanProperty;
+import javafx.beans.property.SimpleBooleanProperty;
+
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * We often need to wait until network and wallet is ready or other combination of startup states.
+ * To avoid those repeated checks for the state or setting of listeners on different domains we provide here a
+ * collection of useful states.
+ */
+@Slf4j
+@Singleton
+public class AppStartupState {
+    // Do not convert to local field as there have been issues observed that the object got GC'ed.
+    private final MonadicBinding<Boolean> p2pNetworkAndWalletInitialized;
+
+    private final BooleanProperty walletAndNetworkReady = new SimpleBooleanProperty();
+    private final BooleanProperty allDomainServicesInitialized = new SimpleBooleanProperty();
+    private final BooleanProperty applicationFullyInitialized = new SimpleBooleanProperty();
+    private final BooleanProperty updatedDataReceived = new SimpleBooleanProperty();
+    private final BooleanProperty isBlockDownloadComplete = new SimpleBooleanProperty();
+    private final BooleanProperty hasSufficientPeersForBroadcast = new SimpleBooleanProperty();
+
+    @Inject
+    public AppStartupState(WalletsSetup walletsSetup, P2PService p2PService) {
+
+        p2PService.addP2PServiceListener(new BootstrapListener() {
+            @Override
+            public void onUpdatedDataReceived() {
+                updatedDataReceived.set(true);
+            }
+        });
+
+        walletsSetup.downloadPercentageProperty().addListener((observable, oldValue, newValue) -> {
+            if (walletsSetup.isDownloadComplete())
+                isBlockDownloadComplete.set(true);
+        });
+
+        walletsSetup.numPeersProperty().addListener((observable, oldValue, newValue) -> {
+            if (walletsSetup.hasSufficientPeersForBroadcast())
+                hasSufficientPeersForBroadcast.set(true);
+        });
+
+        p2pNetworkAndWalletInitialized = EasyBind.combine(updatedDataReceived,
+                isBlockDownloadComplete,
+                hasSufficientPeersForBroadcast,
+                allDomainServicesInitialized,
+                (a, b, c, d) -> {
+                    if (a && b && c) {
```

Comment:
So far not but in future PRs it will be used. The allDomainServicesInitialized might disappear over time or only be used were we need a ordered initialisation, but I think that need to be delegated to the domain where it is used as we did with DaoStartup class. The listeners dont know about the order, but most domains should be independent anyway, and if not it is a sign that there is a parent domain which need to do the initialisation of its children. Of if not there that we miss such a parent daomain.